tests/sentry/backup/test_correctness.py

```diff
@@ -0,0 +1,188 @@
+
+    findings: ComparatorFindings = ComparatorFindings()
+    exp_models: Dict[InstanceID, JSONData] = {}
+    act_models: Dict[InstanceID, JSONData] = {}
+    for model in expect:
+        id = InstanceID(ModelName(model["model"]), PositiveInt(model["pk"]))
+        exp_models[id] = model
+
+    # Ensure that the actual JSON contains no duplicates - we assume that the expected JSON did not.
+    for model in actual:
+        id = InstanceID(ModelName(model["model"]), PositiveInt(model["pk"]))
+        if id in act_models:
+            findings.append(ComparatorFinding(ComparatorName("duplicate_entry"), id))
+        else:
+            act_models[id] = model
+
+    # Report unexpected and missing entries in the actual JSON.
+    act_ids = list(act_models)
+    exp_ids = list(exp_models)
+    for id in act_ids:
+        if id not in exp_ids:
+            del act_models[id]
+            findings.append(ComparatorFinding(ComparatorName("unexpected_entry"), id))
+    for id in exp_ids:
+        if id not in act_ids:
+            del exp_models[id]
+            findings.append(ComparatorFinding(ComparatorName("missing_entry"), id))
+
+    # We only perform custom comparisons and JSON diffs on non-duplicate entries that exist in both
+    # outputs.
+    for id, act in act_models.items():
+        exp = exp_models[id]
+
+        # Try global comparators and comparators applicable for this specific model.
+        for cmp in COMPARATORS[None]:
+            res = cmp(id, exp, act)
+            if res:
+                findings.append(ComparatorFinding(ComparatorName(cmp.__name__), id, res))
+        if id.model in COMPARATORS:
+            for cmp in COMPARATORS[id.model]:
+                res = cmp(id, exp, act)
+                if res:
+                    findings.append(ComparatorFinding(ComparatorName(cmp.__name__), id, res))
+
+        # Finally, perform a diff on the remaining JSON.
+        json_diff = diff(exp["fields"], act["fields"], syntax="explicit", marshal=True)
+        if json_diff:
+            findings.append(
+                ComparatorFinding(ComparatorName("json_diff"), id, json.dumps(json_diff))
+            )
+
+    findings.maybe_assert()
+    return findings
+
+
+def import_then_export(
+    backup_json_file_path: FilePath, silent: bool = True
+) -> Tuple[JSONData, JSONData, ComparatorFindings]:
+    """Test helper that originally imported data correctly matches actual outputted data, and
+    produces a list of reasons why not when it doesn't"""
+
+    with open(backup_json_file_path) as backup_file:
+        contents = backup_file.read()
+        instream = io.StringIO(contents)
+        input = json.load(instream)
+        with in_test_psql_role_override("postgres"):
+            try:
+                instream.seek(0)
+                importer(instream)
+            except IntegrityError as e:
+                assert False, f"Raised an exception {e}"
```

Comment:
this hides the stacktrace -- especially under test this should just raise

Reply:
Done.